### ISSUE TITLE
Option to swap the `softmax`'s backward function with `log_softmax`'s

### ISSUE THREAD

**Issue by mrakgr** [pytorch/pytorch#58822]
## 🚀 Feature

PyTorch has `log_softmax` and `softmax`. I am doing a RL agent and I do not care about calculating the log probabilities, just the regular ones with which to sample from. But I'd like to pass the rewards through the `log_softmax` on the backwards pass. As it happens, the `log_softmax`'s backward function does not require the log probabilities, just the regular ones. 

## Motivation

Optimizing performance, memory use and code cleanliness. 

## Pitch

Extend `softmax` so it takes a flag whether it should act as the `log_softmax` on the backward pass.

## Alternatives

Right now I am going to have to either use the regular `softmax` and `log` it later just to pass the gradients through that, or I am going to have to use the `log_softmax` and then `exp` it to get the actual probability distribution from which to sample from. I am going to go with later as I anticipate it being more numerically stable, but I do not like the either option.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano

**Comment by zou3519** [comment]
Does a custom autograd.Function work for your use case? https://pytorch.org/tutorials/beginner/examples_autograd/two_layer_net_custom_function.html

**Comment by albanD** [comment]
If you want to compute something that is not the actual gradient of the function that is evaluated, you will need to use a custom Function indeed.

**Comment by mrakgr** [comment]
I'd be up for making a custom function, but I am having trouble figuring out where the forward and the backward primitives for the softmax function are. I am having a similar problem with hacking the linear layer. I'd like to hack the backward matmult so it keeps track of input and gradient variances in a moving average, but I am not sure where to start.

When I follow the trail via Intellisense, it points me to `_VariableFunctions.pyi`. Where is the source?

**Comment by albanD** [comment]
I'm afraid you won't be able to copy-paste these functions as they are implemented in c++
But you don't need that with a custom Function as described here: https://pytorch.org/docs/stable/notes/extending.html
You can simply write the formula you want for the backward.
